src/main/java/com/conveyal/datatools/manager/controllers/api/UserController.java

```diff
@@ -145,79 +141,158 @@ private static int getUserCount(Request req, Response res) {
+            String newPermissions = jsonNode.get("permissions").toString();
+            String existingPermissions = searchResult.existingPermissions;
+            // preserve any previously held permissions.
+            String updatedPermissions = (existingPermissions.isEmpty()) ?
+                newPermissions : String.format("%s,%s", existingPermissions, newPermissions);
+            String json = "{ \"app_metadata\": { \"datatools\" : [" + updatedPermissions + "] }}";
+            return updateUser(searchResult.user, searchResult.user.getUser_id(), json, req);
+        }
+
+        // Create new user.
+        String requestBody = String.format("{" +
+                            "\"connection\": \"Username-Password-Authentication\"," +
+                            "\"email\": %s," +
+                            "\"password\": %s," +
+                            "\"app_metadata\": {\"datatools\": [%s] } }"
+                            , jsonNode.get("email"), jsonNode.get("password"), jsonNode.get("permissions"));
         HttpPost createUserRequest = new HttpPost(baseUsersUrl);
         setHeaders(req, createUserRequest);
-
-        JsonNode jsonNode = parseJsonFromBody(req);
-        String json = String.format("{" +
-                "\"connection\": \"Username-Password-Authentication\"," +
-                "\"email\": %s," +
-                "\"password\": %s," +
-                "\"app_metadata\": {\"datatools\": [{\"permissions\": [], \"projects\": [], \"subscriptions\": [], \"client_id\": \"%s\" }] } }",
-                jsonNode.get("email"), jsonNode.get("password"), AUTH0_CLIENT_ID);
-        setRequestEntityUsingJson(createUserRequest, json, req);
-
+        setRequestEntityUsingJson(createUserRequest, requestBody, req);
         return executeRequestAndGetResult(createUserRequest, req);
     }
 
     /**
-     * HTTP endpoint to create new Auth0 user for the application.
+     * Check for an existing Auth0 user based on email address. If the Auth0 search returns a matching email, create
+     * an {@link Auth0UserSearchResult} containing an {@link Auth0UserProfile} and the raw JSON datatools permissions.
+     * If there is no match, return null.
      */
-    private static String createUser(Request req, Response res) {
-        HttpPost createUserRequest = new HttpPost(baseUsersUrl);
-        setHeaders(req, createUserRequest);
+    private static Auth0UserSearchResult checkForExistingAuth0User(Request req, String email) {
+        String query = String.format("email:%s*", email);
+        String searchResult = Auth0Users.getUnrestrictedAuth0Users(query);
+        // Only ever expecting at most one user.
+        List<Auth0UserProfile> users = JsonUtil.getPOJOFromJSONAsList(req, searchResult, Auth0UserProfile.class);
+
+        if (users == null || users.isEmpty()) {
+            return null;
+        } else {
+            Auth0UserSearchResult result = new Auth0UserSearchResult();
```

Comment:
Didn't notice this before, but it seems like some of the below logic should be contained within the `Auth0UserSearchResult` constructor?

Reply:
Updated.